README
Dump perl structures to JavaScript code.

Example:

use Data::JavaScript;

$hash = {'string' => 'Joseph',
         'array' => [qw(0 1 2 3 4 5 6 7 8 9 a b c d e f)],
         'capitals' => {'Sverige' => 'Stockholm',
                        'Norge' => 'Oslo',
                        'Danmark' => 'Koebenhavn'},
         'and' => [[0, 0], [0, 1]],
         'or' => [[0, 1], [1, 1]],
         'xor' => [[0, 1], [1, 0]]};
$hash->{'ref'} = $hash;

print scalar(jsdump("facts", $hash, 31.4e-1));

Output:
                        
var facts = new Array;
facts[0] = new Object;
facts[0].xor = new Array;
facts[0].xor[0] = new Array;
facts[0].xor[0][0] = 0;
facts[0].xor[0][1] = 1;
facts[0].xor[1] = new Array;
facts[0].xor[1][0] = 1;
facts[0].xor[1][1] = 0;
facts[0].or = new Array;
facts[0].or[0] = new Array;
facts[0].or[0][0] = 0;
facts[0].or[0][1] = 1;
facts[0].or[1] = new Array;
facts[0].or[1][0] = 1;
facts[0].or[1][1] = 1;
facts[0].capitals = new Object;
facts[0].capitals.Norge = 'Oslo';
facts[0].capitals.Sverige = 'Stockholm';
facts[0].capitals.Danmark = 'Koebenhavn';
facts[0].and = new Array;
facts[0].and[0] = new Array;
facts[0].and[0][0] = 0;
facts[0].and[0][1] = 0;
facts[0].and[1] = new Array;
facts[0].and[1][0] = 0;
facts[0].and[1][1] = 1;
facts[0].ref = facts[0];
facts[0].array = new Array;
facts[0].array[0] = 0;
facts[0].array[1] = 1;
facts[0].array[2] = 2;
facts[0].array[3] = 3;
facts[0].array[4] = 4;
facts[0].array[5] = 5;
facts[0].array[6] = 6;
facts[0].array[7] = 7;
facts[0].array[8] = 8;
facts[0].array[9] = 9;
facts[0].array[10] = 'a';
facts[0].array[11] = 'b';
facts[0].array[12] = 'c';
facts[0].array[13] = 'd';
facts[0].array[14] = 'e';
facts[0].array[15] = 'f';
facts[0].string = 'Joseph';
facts[1] = 3.14;


CHANGES
Revision history for Perl extension Data::JavaScript.

1.12  Wed Aug 13 22:48:12 EDT 2008
	- Some minor refactoring, including the removal of a dependency on
	  Exporter
	- Fixed a misnumbered test in 1_11, which was intended to be a
	  development release.
	- Escape </script> based on reports of certain stupid browsers
	  ceasing to parse JavaScript upon encountering this string,
	  even in strings.
	- Added explicit license

1_11  Tue Nov 15 14:30:22 EST 2005
	- Touched up documentation
	- Fixed syntax errors for hash key names that are also JS keywords
	  Reported by Kevin J. of Activestate
	- Hash keys are now dumped in alphabetical order, this makes code
	  easier to debug (as suggested by Scott Wessels) and is necessary
	  for thorough testing.
	- Special characters tab, newline, carriage return, ", and \ are
	  now escaped in place instead of as hex: \t, \n, \r, \", \\

1.10  Sun Nov 13 03:50 EST 2005
	- Touched up documentation
	- Explicitly requires Perl 5
	- Previously undocumented, "private", quotemeta is now __quotemeta
	  though you may still import and use it
	- Defaults to JavaScript1.3 compatible code with Unicode and undefined
	  support, specifc JS=>1.2 on the use line for backwards compatbility.
	- Removed unnecessary evaluation of numeric elements, this fixes octal
	  interpretation of zipcodes, etc. Still, zipcodes such as 02137 would
	  be *correctly* evaluated by the client as 1119. However, in keeping
	  with the principle of least surprise any purely numeric string with
	  a leading zero will be treated as a string and quoted. If you really
	  meant to send a literal octal value, you can always eval the var
	  Similarly, engineering and hex notations are passed through to
	  strings as well.

	  Thanks for reports from Javier Arturo Rodriguez, Marius Feraru,
	  Joseph Annino, Alan Fairless, Kevin J.
	- quotemeta has been completely rewritten with support for Unicode

	  Thanks for reports from Slaven Rezic, Marius Feraru, Sean Burke

1.08  Thu Feb 13 09:04:27 EST 2003
	- Touched up the documentation

1.07  Sat Feb  8 12:40:16 EST 2003
        - Added means to control how undef is dumped, suggested by Slaven Rezic
	  In doing so I have removed the undocumented feature that a list
	  was silently transformed into a list reference.
	- Touched up the documentation

****

Ariel Brosh passed away a few months ago. I volunteered to maintain this
module as it was the inspiration for my own Data::JavaScript::LiteObject

http://use.perl.org/article.pl?sid=03/01/12/226258&mode=nested

****

1.04  Thu Jul 19          2001
	- Fixed quoting back, thanks to Garick Hamlin

1.00  Mon Jun 12 21:03:29 2000
	- original version; created by h2xs 1.19



JavaScript.pm
package Data::JavaScript;
require 5;
use vars qw(@EXPORT @EXPORT_OK %OPT $VERSION);
%OPT = (JS=>1.3);
$VERSION = 1.12;

@EXPORT = qw(jsdump hjsdump);
@EXPORT_OK = '__quotemeta';

use strict;
require Encode unless $] < 5.007;

sub import{
  my $package = shift;

  foreach( @_ ){
    if(ref($_) eq 'HASH'){
      $OPT{JS} = $$_{JS} if exists($$_{JS});
      $OPT{UNDEF} = $$_{UNDEF} if exists($$_{UNDEF});
    }
  }
  $OPT{UNDEF} ||=  $OPT{JS} > 1.2 ? 'undefined' : q('');


#  Data::JavaScript->export_to_level(1, grep {!ref($_)} @_);
  #use (); #imports nothing, as package is not supplied
  if( defined $package ){
    no strict 'refs';

    my $caller = caller;
    my @import = grep { ! length ref } @_;

    if( scalar @import ){
      #XXX only :all (@EXPORT, #EXPORT_OK), or subset thereof
    }
    else{
      @import = @EXPORT;
    }
    
    for my $func (@import) {
      *{"$caller\::$func"} = \&$func;
    }
  }
}

sub hjsdump {
    my @res = (qq(<script type="text/javascript" language="JavaScript$OPT{JS}" />),
	       '<!--', &jsdump(@_), '// -->', '</script>');
    wantarray ? @res : join("\n", @res, "");
}

sub jsdump {
    my $sym  = shift;
    return "var $sym;\n" unless (@_);
    my $elem  = shift;
    my $undef = shift;
    my %dict;
    my @res   = __jsdump($sym, $elem, \%dict, $undef);
    $res[0]   = "var " . $res[0];
    wantarray ? @res : join("\n", @res, "");
}

if( $] < 5.007 ){
    eval <<'EO5';
sub __quotemeta {
  local $_ = shift;

  s<([^ \x21-\x5B\x5D-\x7E]+)>{sprintf(join('', '\x%02X' x length$1), unpack'C*',$1)}ge;

  #This is kind of ugly/inconsistent output for munged UTF-8
  #tr won't work because we need the escaped \ for JS output
  s/\\x09/\\t/g;
  s/\\x0A/\\n/g;
  s/\\x0D/\\r/g;
  s/"/\\"/g;
  s/\\x5C/\\\\/g;

  #Escape </script> for stupid browsers that stop parsing
  s%</script>%\\x3C\\x2Fscript\\x3E%g;

  return $_;
}
EO5
}
    else{
	eval<<'EO58';
sub __quotemeta {
  local $_ = shift;
  if( $OPT{JS} >= 1.3 && Encode::is_utf8($_) ){
      s<([\x{0080}-\x{fffd}]+)>{sprintf '\u%0*v4X', '\u', $1}ge;
  }
  
  {
    use bytes;
    s<((?:[^ \x21-\x7E]|(?:\\(?!u)))+)>{sprintf '\x%0*v2X', '\x', $1}ge;
  }

  s/\\x09/\\t/g;
  s/\\x0A/\\n/g;
  s/\\x0D/\\r/g;
  s/"/\\"/g;
  s/\\x5C/\\\\/g;

  s%</script>%\\x3C\\x2Fscript\\x3E%g;

  return $_;
}
EO58
}

sub __jsdump {
    my ($sym, $elem, $dict, $undef) = @_;
    my $ref;

    unless( $ref = ref($elem) ){
      unless( defined($elem) ){
	return "$sym = @{[defined($undef) ? $undef : $OPT{UNDEF}]};";
      }

      #Translated from $Regexp::Common::RE{num}{real}
      if( $elem =~ /^[+-]?(?:(?=\d|\.)\d*(?:\.\d{0,})?)$/ ){
#                                                      (?:[eE][+-]?\d+)?
	  return qq($sym = "$elem";) if $elem =~ /^0\d+$/;
	  return "$sym = $elem;";
      }

      #Fall-back to quoted string
      return qq($sym = ") . __quotemeta($elem) . '";';
    }

    #Circular references
    if ($dict->{$elem}) {
        return "$sym = " . $dict->{$elem} . ";";
    }
    $dict->{$elem} = $sym;

    #isa over ref in case we're given objects
    if( $ref eq 'ARRAY' || UNIVERSAL::isa($elem, 'ARRAY') ){
        my @list = ("$sym = new Array;");
        my $n = 0;
        foreach (@$elem) {
            my $newsym = "$sym\[$n]";
            push(@list, __jsdump($newsym, $_, $dict, $undef));
            $n++;
        }
        return @list;
    }
    elsif(  $ref eq 'HASH' || UNIVERSAL::isa($elem, 'HASH') ){
        my @list = ("$sym = new Object;");
        my ($k, $old_k, $v);
        foreach $k (sort keys %$elem) {
	  $k = __quotemeta($old_k=$k);
	  my $newsym = qq($sym\["$k"]);
	  push(@list, __jsdump($newsym, $elem->{$old_k}, $dict, $undef));
        }
        return @list;
    }
    else{
      return "//Unknown reference: $sym=$ref";
    }
}


1;
__END__

=head1 NAME

Data::JavaScript - Dump perl data structures into JavaScript code

=head1 SYNOPSIS

  use Data::JavaScript;                     # Use defaults
  
  @code =  jsdump('my_array',  $array_ref); # Return array for formatting
  $code =  jsdump('my_object', $hash_ref);  # Return convenient string
  $html = hjsdump('my_stuff',  $reference); # Convenience wrapper

=head1 DESCRIPTION

This module is mainly inteded for CGI programming, when a perl script
generates a page with client side JavaScript code that needs access to
structures created on the server.

It works by creating one line of JavaScript code per datum. Therefore,
structures cannot be created anonymously and need to be assigned to
variables. However, this format enables dumping large structures.

The module can output code for different versions of JavaScript.
It currently supports 1.1, 1.3 and you specify the version on the
C<use> line like so:

  use Data::JavaScript {JS=>1.3};          # The new default
  use Data::JavaScript {JS=>1.1};          # Old (pre module v1.10) format

JavaScript 1.3 contains support for UTF-8 and a native C<undefined> datatype.
Earlier versions support neither, and will default to an empty string C<''>
for undefined values. You may define your own default--for either version--at
compile time by supplying the default value on the C<use> line:

  use Data::JavaScript {JS=>1.1, UNDEF=>'null'};

Other useful values might be C<0>, C<null>, or C<NaN>.

=over

=item jsdump('name', \$reference, [$undef]);

The first argument is required, the name of JavaScript object to create.

The second argument is required, a hashref or arrayref.
Structures can be nested, circular referrencing is supported (experimentally).

The third argument is optional, a scalar whose value is to be used en lieu
of undefined values when dumping a structure.

When called in list context, the function returns a list of lines.
In scalar context, it returns a string.

=item hjsdump('name', \$reference, [$undef]);

hjsdump is identical to jsdump except that it wraps the content in script tags.

=item __quotemeta($str)

Not exported by default, this function escapes non-printable and Unicode
characters to promote playing nice with others.

=back

=head1 CAVEATS

Previously, the module eval'd any data it received that looked like a number;
read: real, hexadecimal, octal, or engineering notations. It now passes all
non-decimal values through as strings. You will need to C<eval> on the client
or server side if you wish to use other notations as numbers. This is meant
to protect people who store ZIP codes with leading 0's.

Unicode support requires perl 5.8 or later. Older perls will gleefully escape
the non-printable portions of any UTF-8 they are fed, likely munging it in
the process as far as JavaScript is concerned. If this turns out to be a
problem and there is sufficient interest it may be possible to hack-in UTF-8
escaping for older perls.

=head1 LICENSE

=over

=item * Thou shalt not claim ownership of unmodified materials.

=item * Thou shalt not claim whole ownership of modified materials.

=item * Thou shalt grant the indemnity of the provider of materials.

=item * Thou shalt use and dispense freely without other restrictions.

=back

Or if you truly insist, you may use and distribute this under ther terms
of Perl itself (GPL and/or Artistic License).

=head1 SEE ALSO

L<Data::JavaScript::LiteObject>, L<Data::JavaScript::Anon>, L<CGI::AJAX|CGI::Ajax>

=head1 AUTHOR

Maintained by Jerrad Pierce <jpierce@cpan.org>

Created by Ariel Brosh <schop cpan.org>.
Inspired by WDDX.pm JavaScript support.

=cut
